cranelift/codegen/src/ir/types.rs

```diff
@@ -22,6 +22,11 @@ use target_lexicon::{PointerWidth, Triple};
 ///
 /// SIMD vector types have power-of-two lanes, up to 256. Lanes can be any int/float/bool type.
 ///
+/// Dynamic SIMD vector types have a power-of-two minimum lane count, which is scaled by a target-defined
```

Comment:
Can you say more about the "minimum lane count" concept -- why it exists, and cases where one might provide different minimum lane counts? And why only {1/2, 1, 2} as factors?

In particular if this comes from a machine restriction, I'm somewhat concerned that we would be baking a specific set of details into the type-system framework forever. I'd rather have a somewhat higher-level notion of dynamic types in CLIF, e.g. `i32xN`; and then when we evaluate the `dyn_scale_...` global value, we can either make whatever guarantees we need for `N` if the value is at the discretion of the ISA backend, or for values loaded from the user, we can place additional restrictions as needed (e.g., a backend provides the embedder a "minimum multiple for dynamic vector size" and the embedder uses this).

Reply:
The minimum lane count comes from the fact that we build a dynamic vector type from a fixed base vector type. The fixed base type provides the minimum lane count.

I added the halving/doubling to try to mimic the functionality of the existing SIMD types, and the ones commonly used, but I don't think it's needed and possibly doesn't make sense. But the main reason was for encoding space...

My problem here was that cranelift has a very large range of SIMD types, which made it hard, for me, to add another group of types with a reasonable range. What I wanted to do was just to use a single bit, along with the existing SIMD types, to express a dynamic vector type but I struggled and failed there.... So I'm very much happy to hear suggestions, I would really prefer to not use this 2-bit system.

Reply:
What if we define dynamic vector types not on top of a base fixed-vector type, but on top of the lane type alone, and then make the dynamic factor (the symbolic `N`) provide a bit width, rather than a multiplier? This is sort of starting from I was getting at with `i32xN` above (rather than `i32x4xN`) but then shifting the meaning of `N` a bit so that we can have the same `N` (symbolically at least, in practice this is all resolved at compile time) for every lane type.

Maybe we need a slightly different notation to make this more clear: perhaps (please suggestion other punctuation, anyone!) `i32/N` (`i32`s over `N` bits). Then we could have

```plain
  N = dyn_width_constant 1024
  dt0 = i32/N
  dt1 = i64/N

  v1 := load.dt0 ...
  v2 = bitcast.dt1 v1
  store.dt2 v2, ...
```

I like this because it removes any notion of "minimum lane count" or some other factor (why `4` in `i32x4xN`, other than the fact that historically we started with 128-bit vectors and we're expanding from there?). We just have the important bits: the lane type (`i32`), and the machine width we're scaling to (`N`). Thoughts?

Reply:
> I'm not sure I follow... for one, with `dyn_width_constant` taking an immediate value, shouldn't we really should just use fixed sizes at the IR level? This doesn't appear to enable anything dynamic.

Ah, so I didn't mean for the `dyn_width_constant` to be the interesting part of that example; perhaps it results from a pass that legalizes to a particular microarchitecture with a known width. The point was to have the global value that defines machine width, as we've had so far; maybe `dyn_width_vl` for a vector-length register or somesuch...

The basic idea I was going for was to have a parameter for the machine width, and say "vector of lane type T up to width `N`". Whereas the `i64x2xN` sort of syntax says "an `N`-times multiple of a base 128-bit vector machine". Just holding a different set of two of the three variables in "machine width = lane width * lanes" constant, I guess.

Said another way, the existing vector types `i8x16`, `i16x8`, `i32x4`, `i64x2` can be written as `i8/128`, `i16/128`, `i32/128`, `i64/128`; and then this is making that width configurable, rather than the multiplier over 128 bits base width configurable.

That said...

> The other problem is that I think we should aim to enable writing clif in dynamic terms without losing too much the functionality of fixed widths - otherwise people have to make a hard choice of whether to use it or not.

I also see the advantages of the `i64x2xN` approach with respect to the existing SIMD types: we're basically "slapping a multiplier on it" and everything else remains the same. The downside is the encoding and from-first-principles comprehensibility of the types.

All else being equal, I think either could work, and an easier migration path is important; and you've done the work to prototype the latter; so `i64x2xN` is totally workable. I won't push the above ideas too hard against it :-)

> I think with your suggestion, we now only have types that use the full width of the register and so we wouldn't be able to support widen and narrow operations?

Ah so here it actually gets interesting: what happens to the {1/2, 1, 2} multipliers if we do two narrow or widen ops in a row?

Reply:
> Ah so here it actually gets interesting: what happens to the {1/2, 1, 2} multipliers if we do two narrow or widen ops in a row?

Everything breaks :)

Although it would also break for fixed vectors after a certain amount of doubling too. We're never going to be able to support everything with a small, fixed, number of types, but I would really like dynamic types to have the same limitations as their fixed-width counterparts. I will look at the encoding again, but I think this is tough with how many vector types already exist. I take it that moving to a u16 is out of the question...?

I'll also prototype support for widen and narrow, because now I'm wondering how that will actually work... I assume we shouldn't support an implicit result type, unless that type has already been created for function?

Reply:
I think it could be worthwhile to run some quick experiments with `Type` as a `u16` instead. It may pessimize a tiny bit where we have a `Vec<Type>` rather than a `Type` as a field (in the latter case we probably have some slack with padding) but if it's in the noise or say 0.1% or something, it seems like it would make this work substantially easier...

> I'll also prototype support for widen and narrow, because now I'm wondering how that will actually work... I assume we shouldn't support an implicit result type, unless that type has already been created for function?

Yes, actually, now that I think about it more, this seems right (and like a small difference from existing fixed SIMD types): one will need to explicitly declare the types on both sides. I'm curious to see how that will work out wrt our instruction constraint system; I don't have enough details paged in right now to say for sure if there will be any problems...

Reply:
Sounds good to me!

Reply:
@cfallin What should I be running for looking at to get the metrics that you're interested in?

Reply:
Any sort of profiling (even just a `perf stat ...`) on a compilation of a large module (`benchmarks-next/spidermonkey/benchmark.wasm` from the Sightglass repo?) should be enough to see if making `Type` a `u16` has any impact, I think. Though given the above I *expect* it shouldn't and would be surprised if we see anything more than a blip...

Reply:
Taking the example that you suggested, compilation is 0.7% slower when using u16. This is taking the mean over three runs. I'm using a laptop... so not the best benchmarking setup, but hopefully this sounds like a low enough impact?

Reply:
Hmm, that's actually a bit higher than I was expecting. Wins from significant efforts can sometimes be 5-10% (and e.g. RA2 with ~6 months of work was 10-20% compile time reduction), so throwing away 0.7% here should be avoided if we can find a way around it at all.

I suspect it might be [this](https://github.com/bytecodealliance/wasmtime/blob/7e0bb465d0205fd7942a48dce931a8221dccabb8/cranelift/codegen/src/machinst/vcode.rs#L63) `Vec<Type>` in the `VCode`. Could you run another experiment where you keep `Type` at `u8` but just inflate that to `u16` (maybe make it a `Vec<(Type, u8)>` and the minimal hacks to stuff the extra `u8` in)?

If it turns out to be that `Vec`, we could keep it a `Vec<u8>`, reserve one `u8` value for `>= 0xff`, and keep a `FxHashMap<VReg, Type>` sparse map for the "out of bounds" types, if that makes sense.

Otherwise, would you be willing to hunt down the perf delta a bit and see if it's coming from somewhere else?

Reply:
> Could you run another experiment where you keep Type at u8 but just inflate that to u16 (maybe make it a Vec<(Type, u8)>

This actually gave a small speedup instead!

> we actually just stuff the Type into the ValueData [here](https://github.com/bytecodealliance/wasmtime/blob/72c6b79e9c96c75b7fa31e4d84ece5145e4a9aa3/cranelift/codegen/src/ir/dfg.rs#L431) so we almost certainly have the padding to absorb a u16 instead

Unfortunately, this doesn't appear to be the case, at least this appears to be where the regression is coming from. perf showed `DataFlowGraph::value_def` taking a longer time, so I converted `ty` into `(Type, u8)` in the `ValueData` enum and saw a 1.4% slow down. Setting `repr(align(8))` fixed this regression, but not for the original case of `Type(u16)`....

However, using `num: usize` for `ValueData::Inst` and `ValueData::Param` reduces the regression to 0.19% when using u16.

Reply:
Ah, I think I missed that the enum dscriminant in `ValueData` is an implicit `u8`, so right now it fits in `u64` but making `Type` a `u16` inflates that (probably at least to 12 bytes if not 16?).

I have some thoughts on bitpacking and giving `Type` 14 bits instead; I'll do a PR and do some measurements, and get this in as a separate refactor if it works to pave the way for your work :-)

Reply:
@sparker-arm I took a look at this just now in #4269. The impact on compiling SpiderMonkey.wasm was actually 4.5%, which is quite serious and so we definitely need to do something here -- the bitpacking in that PR gets it down to 1.0% but I'm hoping we can do better (ideally I wouldn't want to pay more than 0.1% for this, given that compile-time wins are generally hard to come by).

Reply:
So, I re-ran my experiments and looking at reported time, instead of raw cycles, and I see u16 causing a 1.1% or 1.3% slowdown, depending on whether I use the mean or minimum values. (I don't understand why cycles and time are so different...) And I'm a bit concerned that your results are significantly different than mine. What are machine you running on? And are you running sightglass in a way to mimic a 'real-life' scenario, with the commands `--iterations-per-process 10 --processes 2`?

To add to my confusion, when I benchmark your 14-bit patch I confirm your results, with a 0.78% slowdown...

Reply:
Results for me were on a desktop Ryzen 3900X 12-core system. The iteration count and process count were a tradeoff to get some results and not wait forever, but all else equal, larger counts are better (it lets Sightglass get a tighter confidence interval). If run with just two engines rather than three it'll do a statistical-significance test too.

Compilation input can make a big difference in slowdowns observed (e.g. if functions get big enough for IR data structures to exceed L1), are you testing with SpiderMonkey.wasm or something else?